proposals/0047-rerouting.md

```diff
@@ -0,0 +1,186 @@
+
+```js
+
+export const GET = (context) => {
+    return context.reroute("/fr/hello")
+}
+```
+### Middleware usage
+
+Other than exposing the `reroute` function via `context`, the signature of the `next` function will change to accept the same payload of the `reroute` function:
+
+```js
+export const onRequest = (context, next) => {
+    if (context.url.startsWith("/fr/salut")) {
+        return next("/fr/hello")
+    }
+    return next()
+}
+```
+
+The `next` function will have this new signature, and it will be backwards compatible with the existing signature:
+
+```ts
+type MiddlewareNext = (payload?: Reroute) => Promise<Response>;
+```
+
+
+### Results for static and prerendered pages
+
+In SSG, rerouting to a page will result in the compilation of a page that has the contents of a different page. 
+
+If we have two pages, `src/pages/about.astro` and `src/pages/contact.astro`, if the `contact.astro` will reroute to `/about` using the rerouting strategy, it's contents will be the same as the `about.astro` component.
+
+```html
+<!-- src/pages/about.astro-->
+<html>
+  <head>
+    <title>Astro</title>
+  </head>
+  <body>About</body>
+</html>
+```
+
+```astro
+---
+// src/pages/contact.astro
+
+Astro.reroute("/about");
+---
+```
+The page in `dist/contact/index.html` will contain:
+```html
+<html>
+  <head>
+    <title>Astro</title>
+  </head>
+  <body>About</body>
+</html>
+```
+
+### Results for on-demand pages
+
+In SSR, rerouting to a page will result in rendering the contents of said page. The example provided for the static pages applies for on-demand pages too.
+
+
+## Expectations upon rerouting
+
+When a user triggers a rerouting to another URL/route:
+- The middleware is **NOT** triggered **again** upon rerouting.
+- If the rerouted route doesn't match any of the existing routes, a 404 `Response` is returned. The middleware **is triggered** for the 404 (it keeps the existing behaviour)
+- The rerouted route will render the first page/route that is matched, among the list of [sorted routes](https://docs.astro.build/en/guides/routing/#route-priority-order).
+- Injected routes should be eligible from said matching.
+
+## Adapters
+
+I don't envision any major blocker for our current adapters. 
+
+I envision some potential changes for edge middleware, where the `reroute` function should be overridden with a platform specific function.
+ 
+
+# Testing Strategy
+
+- Integration testing
+- Manual testing
+
+# Drawbacks
+
+The current API is a blocker for `functionPerRoute`. Since the rerouting isn't statically analysable, it isn't possible to tell serverless computing to reroute a request to another function/lambda. 
+
+# Alternatives
+
+I considered a static approach like SvelteKit and Next.js. While this approach would allow to achieve `functionPerRoute` support, we risk to litter the configuration with many entries. Plus, a static approach doesn't allow enough flexibility. 
```

Comment:
One alternative that we discussed was making this purely a middleware feature doing something like `next({ request: new Request('/other')... })`. I would add that and discuss the tradeoffs.